### PR TITLE
Ppc64le multiarch

### PR DESCRIPTION
**Pull request by sumitd2**
Draft PR, ppc64le support with multiarch images

### CODE REVIEW

File: builder/install/50-luajit.sh [hunk 1]
```diff
@@ -2,8 +2,14 @@
...
-
-cp COPYRIGHT "${LICENSE_DIR}/LuaJIT-${LUAJIT_VERSION}"
+if [ "$(arch)" = "x86_64" ]; then
+    cd third_party/luajit
+    cp COPYRIGHT "${LICENSE_DIR}/LuaJIT-${LUAJIT_VERSION}"
+else
+    cd third_party
+    git clone https://github.com/openresty/luajit2.git
+    cd luajit2
+    git checkout v"${LUAJIT2_VERSION}"
```
Comment (yselkowitz) [inline builder/install/50-luajit.sh:12]: luajit2 needs to be imported properly as a submodule.

File: integration-tests/collector_manager.go [hunk 2]
```diff
@@ -60,14 +60,15 @@ func NewCollectorManager(e Executor, name string) *collectorManager {
 	}
 
 	collectorImage := ReadEnvVar("COLLECTOR_IMAGE")
+	grpcServerImage := "quay.io/rhacs-eng/grpc-server:3.70.x-468-g5d733366eb"
```
Comment (yselkowitz) [inline integration-tests/collector_manager.go:63]: Ditto for this file.

File: integration-tests/container/benchmark/Dockerfile [hunk 3]
```diff
@@ -1,5 +1,13 @@
-FROM phoronix/pts
+FROM --platform=linux/amd64 phoronix/pts as stage-amd64
+FROM --platform=linux/ppc64le quay.io/rhacs-eng/phoronix-pts-ppc64le as stage-ppc64le
```
Comment (yselkowitz) [inline integration-tests/container/benchmark/Dockerfile:2]: Is this image already available?

File: integration-tests/container/benchmark/Dockerfile [hunk 4]
```diff
@@ -1,5 +1,13 @@
...
 
+# Declare TARGETARCH to make it available
+ARG TARGETARCH
+
+# Select final stage based on TARGETARCH ARG
+FROM stage-${TARGETARCH} as final
```
Comment (yselkowitz) [inline integration-tests/container/benchmark/Dockerfile:8]: If this is just for CI then this should work there, but tricks such as this will not work for downstream builds.

File: integration-tests/container/perf/Dockerfile [hunk 5]
```diff
@@ -1,5 +1,13 @@
-FROM alpine:3.12 AS base
+FROM --platform=linux/amd64 alpine:3.12 AS stage-amd64
+FROM --platform=linux/ppc64le ppc64le/alpine:3.13 AS stage-ppc64le
```
Comment (yselkowitz) [inline integration-tests/container/perf/Dockerfile:2]: `alpine` is already a multi-arch manifest-listed image, so no need to jump through such hoops in this case.  If for whatever reason you need/want 3.13, probably just best to use it across the board.

File: integration-tests/container/perf/Dockerfile.bpftrace [hunk 6]
```diff
@@ -1,4 +1,13 @@
-FROM quay.io/iovisor/bpftrace:latest
+FROM --platform=linux/amd64 quay.io/iovisor/bpftrace:latest AS stage-amd64
+
+FROM --platform=linux/ppc64le ubuntu:focal AS stage-ppc64le
+RUN apt-get update -y && apt-get install -y bpftrace && bpftrace -V
```
Comment (yselkowitz) [inline integration-tests/container/perf/Dockerfile.bpftrace:4]: Looks like the original image also installs `xz-utils`, but either way, if this is all it is, why not just start with a base image and install bpftrace on all arches?

File: integration-tests/container/schedule-curls/Dockerfile [hunk 7]
```diff
@@ -1,3 +1,10 @@
-FROM pstauffer/curl:latest
+FROM --platform=linux/amd64 pstauffer/curl:latest as stage-amd64
+FROM --platform=linux/ppc64le quay.io/rhacs-eng/pstauffer-curl-ppc64le:latest as stage-ppc64le
```
Comment (yselkowitz) [inline integration-tests/container/schedule-curls/Dockerfile:2]: Why not just `docker.io/curlimages/curl` which is both multi-arch and current?

File: integration-tests/integration_test.go [hunk 8]
```diff
@@ -427,7 +427,8 @@ func (s *RepeatedNetworkFlowTestSuite) SetupSuite() {
 	err = s.collector.Launch()
 	s.Require().NoError(err)
 
-	scheduled_curls_image := qaImage("quay.io/rhacs-eng/qa", "collector-schedule-curls");
+	image := "quay.io/rhacs-eng/qa"
```
Comment (yselkowitz) [inline integration-tests/integration_test.go:430]: The cumulative changes in this file also end up being just cosmetic; drop them entirely.

File: kernel-modules/build/Dockerfile [hunk 9]
```diff
@@ -18,6 +20,14 @@ RUN apt-get update \
       pkg-config \
  ;
 
+RUN git clone https://github.com/google/googletest.git && \
```
Comment (yselkowitz) [inline kernel-modules/build/Dockerfile:23]: Seems like this should also be properly submodule'd.

File: kernel-modules/build/Dockerfile.fc36 [hunk 10]
```diff
@@ -16,10 +18,19 @@ RUN dnf update -y || true; \
         kmod \
         wget \
         golang-go \
+        dwarves \
         pkg-config || true ; \
     mkdir -p /output; \
     dnf autoremove -y
 
+RUN git clone https://github.com/google/googletest.git && \
```
Comment (yselkowitz) [inline kernel-modules/build/Dockerfile.fc36:26]: Ditto.

File: reload/missing-kernel-objects.sh [hunk 11]
```diff
@@ -22,7 +22,6 @@ output_dir="$3"
 mkdir -p "${output_dir}/${collector_version}"
 
 image="quay.io/rhacs-eng/collector:${collector_version}-latest"
-
```
Comment (yselkowitz) [inline reload/missing-kernel-objects.sh:25]: Cosmetic change only.

File: scripts/push-as-manifest-list.sh [hunk 12]
```diff
@@ -14,7 +14,12 @@ image="$1"
 [[ -n "$image" ]] || die "No image specified"
 [[ "$image" == *:* ]] || die "Must specify a tagged image reference when using this script"
 
-arch_image="${image}-amd64"
+ARCH=$(arch)
+if [ "$ARCH" = "x86_64" ]; then
+    arch_image="${image}-amd64"
+else
+    arch_image="${image}-$ARCH"
+fi
```
Comment (yselkowitz) [inline scripts/push-as-manifest-list.sh:22]: Please think long-term here.  ARM also has such a naming disparity (uname: `aarch64`, goarch: `arm64`).

File: collector/CMakeLists.txt [hunk 13]
```diff
@@ -93,6 +93,11 @@ add_test(collector-tests runUnitTests)
 # Falco Wrapper Library
 set(BUILD_DRIVER OFF CACHE BOOL "Build the driver on Linux" FORCE)
 set(USE_BUNDLED_DEPS OFF CACHE BOOL "Enable bundled dependencies instead of using the system ones" FORCE)
+if(${CMAKE_SYSTEM_PROCESSOR} STREQUAL "x86_64")
+	set(USE_BUNDLED_LUAJIT OFF CACHE BOOL "Disable building of the bundled LuaJIT" FORCE)
+else()
+	set(USE_BUNDLED_LUAJIT ON CACHE BOOL "Enable building of the bundled LuaJIT" FORCE)
+endif()
```
Comment (yselkowitz) [inline collector/CMakeLists.txt:100]: Hard to tell from the context; does this change the default on x86?  Where is this variable used?

**Review by yselkowitz — CHANGES_REQUESTED** [review]
A few general points:

- `docker buildx build` cannot be simply followed up by `docker push` or `docker run` etc.; the docker daemon cannot directly access `buildx build` images.  If you're just building the image to see if it will build on multiple arches, then fine, but if you're planning on pushing or running the images, then you need to adapt those steps accordingly as well.
- `docker buildx` and `TARGETPLATFORM` hacks are fine for upstream/CI but not for downstream; it doesn't seem like these dockerfiles could or would be used downstream anyway, but just checking.
- There are a few files noted below which, once squashed, end up with only cosmetic changes; just remove those entirely.
- What about s390x?

**Review by ovalenti — COMMENTED** [review]
I must admit that the project strongly assumes an x86_64 arch and will benefit from being "enabled" on other platforms. So overall, thanks for the work !

On a strategic level, I think that it would be cleaner to add parameters to the build than to add tests about the architecture as part of the build process.

Using the current approach, for instance, it would become quite complex to target several ppc64 distros in the future, and each arch/distro added will require to do these changes again.

In conclusion, I wish we could keep arch/distro specific stuff only inside the code creating CI/CD pipelines, providing as many parameters as required to handle the specifics of each build (like the base image for the builder, for instance).

File: .gitmodules [hunk 14]
```diff
@@ -50,3 +50,7 @@
 	path = third_party/valijson
 	url = https://github.com/tristanpenman/valijson.git
 	branch = master
+[submodule "third_party/luajit2"]
+	path = third_party/luajit2
+	url = https://github.com/openresty/luajit2.git
+	branch = master
```
[the remark this answers is not in the document]
Comment (sumitd2) [reply .gitmodules:56]: luajit does not support ppc64le

File: integration-tests/container/benchmark/Dockerfile (hunk 3 above)
Comment (sumitd2) [reply integration-tests/container/benchmark/Dockerfile:2]: This image will have to be build once for power.

File: kernel-modules/build/Dockerfile (hunk 9 above)
Comment (sumitd2) [reply kernel-modules/build/Dockerfile:23]: googletest is already a submodule of this repo. I am not sure what is the proper way to get it into this dockerfile.

File: kernel-modules/build/Dockerfile.fc36 (hunk 10 above)
Comment (sumitd2) [reply kernel-modules/build/Dockerfile.fc36:26]: googletest is already a submodule of this repo. I am not sure what is the proper way to get it into this dockerfile.

File: builder/install/50-luajit.sh (hunk 1 above)
Comment (sumitd2) [reply builder/install/50-luajit.sh]: Done

File: integration-tests/benchmark_test.go [hunk 15]
```diff
@@ -37,6 +37,7 @@ func (b *BenchmarkTestSuiteBase) StartPerfTools() {
 	perf := ReadEnvVar("COLLECTOR_PERF_COMMAND")
 	bpftrace := ReadEnvVar("COLLECTOR_BPFTRACE_COMMAND")
 	bcc := ReadEnvVar("COLLECTOR_BCC_COMMAND")
+	image := "quay.io/rhacs-eng/collector-performance"
```
[the remark this answers is not in the document]
Comment (sumitd2) [reply integration-tests/benchmark_test.go]: Done

File: integration-tests/collector_manager.go (hunk 2 above)
Comment (sumitd2) [reply integration-tests/collector_manager.go]: Done

File: integration-tests/container/benchmark/Dockerfile (hunk 4 above)
Comment (sumitd2) [reply integration-tests/container/benchmark/Dockerfile:8]: Not sure why this wont work?

File: reload/missing-kernel-objects.sh (hunk 11 above)
Comment (sumitd2) [reply reload/missing-kernel-objects.sh]: Done

File: scripts/push-as-manifest-list.sh (hunk 12 above)
Comment (sumitd2) [reply scripts/push-as-manifest-list.sh]: Done

File: collector/CMakeLists.txt (hunk 13 above)
Comment (sumitd2) [reply collector/CMakeLists.txt]: Done

File: .gitmodules (hunk 14 above)
[the remark this answers is not in the document]
Comment (ovalenti) [reply .gitmodules:56]: If luajit2 works for ppc, then maybe we can use it for all archs ?

[the remark this answers is not in the document]
Comment (sumitd2) [reply .gitmodules:56]: luajit2 is not actively maintained, plus we did not want to modify the current x86 functionality

**Comment by sumitd2** [comment]
> * `docker buildx build` cannot be simply followed up by `docker push` or `docker run` etc.; the docker daemon cannot directly access `buildx build` images.  If you're just building the image to see if it will build on multiple arches, then fine, but if you're planning on pushing or running the images, then you need to adapt those steps accordingly as well.
**We can use neither buildx build --load (doesnt work with multiarch images), nor --push (we cant push after every build). Looks like this PR is a waste.**
> * `docker buildx` and `TARGETPLATFORM` hacks are fine for upstream/CI but not for downstream; it doesn't seem like these dockerfiles could or would be used downstream anyway, but just checking.
**No sure why.**
> * There are a few files noted below which, once squashed, end up with only cosmetic changes; just remove those entirely.
**Done**
> * What about s390x?
**That will be a separate PR from a different team.**

File: .gitmodules (hunk 14 above)
[the remark this answers is not in the document]
Comment (yselkowitz) [reply .gitmodules:56]: luajit2 seems to be well maintained?

File: kernel-modules/build/Dockerfile (hunk 9 above)
Comment (yselkowitz) [reply kernel-modules/build/Dockerfile:23]: with `WORKDIR` and `COPY` directives

File: kernel-modules/build/Dockerfile.fc36 (hunk 10 above)
Comment (yselkowitz) [reply kernel-modules/build/Dockerfile.fc36:26]: with `WORKDIR` and `COPY` directives

File: .gitmodules (hunk 14 above)
[the remark this answers is not in the document]
Comment (sumitd2) [reply .gitmodules:52]: you are right, i confused luajit2 with moonjit, which is another fork. Now using it for both x86 and power.

**Comment by ovalenti** [comment]
> On a strategic level, I think that it would be cleaner to add parameters to the build than to add tests about the architecture as part of the build process.
> ...
> In conclusion, I wish we could keep arch/distro specific stuff only inside the code creating CI/CD pipelines, providing as many parameters as required to handle the specifics of each build (like the base image for the builder, for instance).

@sumitd2 , I would love to hear about your opinion on this subject, as it appears that your are pursuing in the exact same direction as before.

**Comment by sumitd2** [comment]
@ovalenti 
Hi Olivier, apologies for the late reply. I am still trying to comprehend your suggestions. I am responsible for porting the code to power, and don't know much about the code or the functionality it implements. I have asked a Redhat team member (@robbycochran) to analyse your post and post a suitable reply. I am taking the same direction in the meanwhile as we are in a bit of a hurry for the first release, I just thought your suggestions could possibly be implemented in the second one. Thank you.

File: kernel-modules/build/Dockerfile (hunk 9 above)
Comment (sumitd2) [reply kernel-modules/build/Dockerfile]: Done

File: kernel-modules/build/Dockerfile.fc36 (hunk 10 above)
Comment (sumitd2) [reply kernel-modules/build/Dockerfile.fc36]: Done

File: integration-tests/benchmark_test.go [hunk 16]
```diff
@@ -37,6 +37,11 @@ func (b *BenchmarkTestSuiteBase) StartPerfTools() {
...
+
+	if runtime.GOARCH != "amd64" {
+		image = "quay.io/rhacs-eng/" + runtime.GOARCH + "/collector-performance"
```
Comment (hbrueckner) [inline integration-tests/benchmark_test.go]: What about specifying this image as environment variables as well? This might be beneficial for self-built images or would further abstract arch specifics for now... I would guess that in future, this container image will be built w/ manifest-list / multi-arch support.

File: integration-tests/benchmark_test.go [hunk 17]
```diff
@@ -72,7 +77,12 @@ func (b *BenchmarkTestSuiteBase) StartPerfContainer(name string, image string, a
 }
 
 func (b *BenchmarkTestSuiteBase) RunInitContainer() {
-	init_image := qaImage("quay.io/rhacs-eng/collector-performance", "init")
+	image := "quay.io/rhacs-eng/collector-performance"
+
+	if runtime.GOARCH != "amd64" {
```
Comment (hbrueckner) [inline integration-tests/benchmark_test.go]: same here .. see above.

File: utilities/PerformanceTesting/start-secured-cluster.sh [hunk 18]
```diff
@@ -9,7 +9,6 @@ else
     collector_image_registry=${2:-quay.io/rhacs-eng/$ARCH}
 fi
 
-
```
Comment (hbrueckner) [inline utilities/PerformanceTesting/start-secured-cluster.sh]: cosmetic

Comment (sumitd2) [reply utilities/PerformanceTesting/start-secured-cluster.sh]: Already fixed. Cant see this in the "Files Changed" tab.

File: integration-tests/benchmark_test.go (hunk 16 above)
Comment (sumitd2) [reply integration-tests/benchmark_test.go]: arch specifics have already been removed from here. Cant see this in the "Files Changed" tab.

File: integration-tests/benchmark_test.go (hunk 17 above)
Comment (sumitd2) [reply integration-tests/benchmark_test.go]: arch specifics have already been removed from here. Cant see this in the "Files Changed" tab.
